src/NumericalAlgorithms/Spectral/ComplexDataView.cpp

```diff
@@ -0,0 +1,274 @@
+  complex_view_.set_data_ref(start, size);
+}
+
+// RealsThenImags constructors cause an allocation and a copy for each
+// component, and a copy for each component. If it becomes a performance
+// concern, the `real_slice_` and `imag_slice_` could be made as non-owning
+// DataVectors which reference a larger block of memory allocated only once.
+
+// clang-tidy: This class uses manual memory management deliberately,
+// so silence complaints about pointer math and casts
+template <>
+ComplexDataView<ComplexRepresentation::RealsThenImags>::ComplexDataView(
+    gsl::not_null<ComplexDataVector*> vector, size_t size,
+    size_t offset) noexcept
+    : size_{size},
+      real_slices_up_to_date_{true},
+      complex_view_up_to_date_{true},
+      complex_view_{} {
+  complex_view_.set_data_ref(vector->data() + offset, size_);
+  real_slice_ = real(complex_view_);
+  imag_slice_ = imag(complex_view_);
+  data_real_ = real_slice_.data();
+  data_imag_ = imag_slice_.data();
+}
+
+template <>
+ComplexDataView<ComplexRepresentation::RealsThenImags>::ComplexDataView(
+    std::complex<double>* start, size_t size) noexcept
+    : size_{size},
+      real_slices_up_to_date_{true},
+      complex_view_up_to_date_{true},
+      complex_view_{} {
+  complex_view_.set_data_ref(start, size_);
+  real_slice_ = real(complex_view_);
+  imag_slice_ = imag(complex_view_);
+  data_real_ = real_slice_.data();
+  data_imag_ = imag_slice_.data();
+}
+
+// For Interleaved views, assigning individual components is comparatively
+// slow, and is currently implemented as a loop over copies.
+
+// clang-tidy: This class uses manual memory management deliberately,
+// so silence complaints about pointer math and casts
+template <>
+ComplexDataView<ComplexRepresentation::Interleaved>&
+ComplexDataView<ComplexRepresentation::Interleaved>::assign_real(
+    const DataVector& vector) noexcept {
+  sizes_assert(vector, size_);
+  for (size_t i = 0; i < size_; i++) {
+    data_real_[stride_ * i] = vector[i];  // NOLINT
+  }
+  real_slices_up_to_date_ = false;
+  return *this;
+}
+
+// clang-tidy: This class uses manual memory management deliberately,
+// so silence complaints about pointer math and casts
+template <>
+ComplexDataView<ComplexRepresentation::Interleaved>&
+ComplexDataView<ComplexRepresentation::Interleaved>::assign_imag(
+    const DataVector& vector) noexcept {
+  sizes_assert(vector, size_);
+  for (size_t i = 0; i < size_; i++) {
+    data_imag_[stride_ * i] = vector[i];  // NOLINT
+  }
+  real_slices_up_to_date_ = false;
+  return *this;
+}
+
+// For RealThenImags views, the assignment to individual components assigns
+// directly to the internal DataVectors of the separate components, which is
+// comparatively fast.
+template <>
+ComplexDataView<ComplexRepresentation::RealsThenImags>&
+ComplexDataView<ComplexRepresentation::RealsThenImags>::assign_real(
+    const DataVector& vector) noexcept {
+  sizes_assert(vector, size_);
```

Comment:
Maybe, I was less convinced about what would happen in the case of RealsThenImags. We've generally been checking for self-assignment (like what you did in your fixup :) ). Would you mind testing the self-assignment since errors in this class could cause serious headaches in the future for someone using libsharp (I guess you, initially, but still :) )?